packages/connect/src/api/firmware/verifyAuthenticityProof.ts

```diff
@@ -0,0 +1,165 @@
+    const SubtleCrypto = typeof window !== 'undefined' ? window.crypto.subtle : crypto.subtle;
+    if (!SubtleCrypto) {
+        throw new Error('SubtleCrypto not supported');
+    }
+
+    // get ECDSA P-256 (secp256r1) key from RAW key
+    const ecPubKey = await SubtleCrypto.importKey(
+        'raw',
+        rawKey,
+        { name: 'ECDSA', namedCurve: 'P-256' },
+        true,
+        ['verify'],
+    );
+
+    // export ECDSA key as spki
+    const spkiPubKey = await SubtleCrypto.exportKey('spki', ecPubKey);
+
+    // create PEM from spki
+    const key = `-----BEGIN PUBLIC KEY-----\n${Buffer.from(spkiPubKey).toString(
+        'base64',
+    )}\n-----END PUBLIC KEY-----`;
+
+    // verify using PEM key
+    return signer.verify({ key }, Buffer.from(signature));
+};
+
+interface AuthenticityProofData extends PROTO.AuthenticityProof {
+    challenge: Buffer;
+    config: DeviceAuthenticityConfig;
+    deviceModel: keyof typeof PROTO.DeviceModelInternal; // Device.features.internal_model
+}
+
+export const getRandomChallenge = () => crypto.randomBytes(32);
+
+export const verifyAuthenticityProof = async ({
+    ca_certificate,
+    device_certificate,
+    signature,
+    challenge,
+    config,
+    deviceModel,
+}: AuthenticityProofData): Promise<AuthenticateDeviceResponse> => {
+    const modelConfig = config[deviceModel];
+    if (!modelConfig) {
+        throw new Error(`Pubkeys for ${deviceModel} not found in config`);
+    }
+    const { rootPubKeys, caPubKeys } = modelConfig;
+
+    // 1. parse x509 CA certificate from AuthenticityProof
+    const caCert = parseCertificate(new Uint8Array(Buffer.from(ca_certificate, 'hex')));
+    const caPubKey = Buffer.from(caCert.tbsCertificate.subjectPublicKeyInfo.bits.bytes).toString(
+        'hex',
+    );
+
+    // 2. parse x509 DEVICE certificate from AuthenticityProof
+    const deviceCert = parseCertificate(new Uint8Array(Buffer.from(device_certificate, 'hex')));
+
+    // 3. validate that CA certificate was created using one of rootPubkeys
+    const isCertSignedByRootPubkey = await Promise.all(
+        rootPubKeys.map(rootPubKey =>
+            verifySignature(
+                Buffer.from(rootPubKey, 'hex'),
+                caCert.tbsCertificate.asn1.raw,
+                caCert.signatureValue.bits.bytes,
+            ),
+        ),
+    );
+
+    const isCaCertValid = isCertSignedByRootPubkey.some(r => !!r);
+    if (!isCaCertValid) {
+        const outdated =
+            new Date(config.timestamp).getTime() < caCert.tbsCertificate.validity.from.getTime();
+        return {
+            valid: false,
+            caPubKey,
+            error: outdated ? 'OUTDATED_CONFIG' : 'ROOT_PUBKEY_NOT_FOUND',
+        };
+    }
+
+    // 4. validate DEVICE certificate subject (Trezor features internal_model)
+    const [subject] = deviceCert.tbsCertificate.subject;
+    if (!subject.parameters || subject.algorithm !== '2.5.4.3') {
```

Comment:
would be nice to add link in comment https://www.alvestrand.no/objectid/2.5.4.3.html

Reply:
done